We will need to write the generic algorithm
a) What does an individual look like, representing a complete maintenance schedule? 
b) What is our selection strategy?
c) What is the crossover method?
d) If we use, what is the mutation method?
e) How do we feel with infeasible individuals after crossover/mutation?
f) What are your population size, your crossover and mutation probabilities, and any other
design choices and hyperparameters you might need?


The algorithm should be suitable to solve the problem, and should be clearly described. It should be able
to accept the predicted **Remaining Useful Lifetime (RULs)** for all engines as input, and output a valid maintenance solution for it. 

Such a solution should provide: 
(i) a list of all machines, indicating for any maintained machine the type of the team doing the maintenance, start-date, and end-date of the maintenance, as well as the penalty costs incurred by that machine;

(ii) the total penalty costs. It is not important here what the concrete cost of such a solution is, it just should be feasible. Make sure your solution format is easy to read and understand

**Input**: RUL predictions  (100 entries/rows)

**Output**: a list of all machines, indicating for any maintained machine the type of the team doing the maintenance, start-date, and end-date of the maintenance, as well as the penalty costs incurred by that machine; (ii) the total penalty costs. (100 entries/rows)

## Read the predicted RUL

In [12]:
import os
import pandas as pd

# Get the current working directory
current_directory = os.getcwd()

# Construct the relative path to prediction RUL file
rul_filename = "RUL_consultancy_predictions_A3-2.csv"
rul_path = os.path.join(current_directory, rul_filename)

# Read the Excel file
rul_df = pd.read_csv(rul_path)
print(rul_df)

    RUL;id
0    135;1
1    125;2
2     63;3
3    100;4
4    103;5
..     ...
95  140;96
96  109;97
97   87;98
98  127;99
99  24;100

[100 rows x 1 columns]


## Lecture 6 from Hendrik Baier ;lecture 6 - evolutionary algorithms 
## First step in Genetic Algorithm: Population

####  Who is an individual? I think it is one machine with its maintenance plan
#### The whole population is all the machines with their maintenance plan?


So the id from the RUL is the id of the engine, right? If we take the 

RUL;id
135;1
125;2
...

Engine number 1 has 135 days left until it needs to be maintained and engine 2 has 125 days ect.

In [1]:
# Define constants
NUM_ENGINES = len(rul_df)
NUM_TEAMS_A = 2
NUM_TEAMS_B = 2
PLANNING_HORIZON = 30
MAX_DAILY_COST = 250

# Define maintenance times for teams A and B
maintenance_times_a = [4 if i < 20 else 3 if 20 <= i < 55 else 2 if 55 <= i < 80 else 8 for i in range(1, NUM_ENGINES + 1)]
maintenance_times_b = [time_a + 1 if i < 25 else time_a + 2 if 25 <= i < 70 else time_a + 1 for i, time_a in enumerate(maintenance_times_a, start=1)]

# Define engine costs
engine_costs = [4 if i < 21 else 3 if 21 <= i < 31 else 2 if 31 <= i < 46 else 5 if 46 <= i < 81 else 6 for i in range(1, NUM_ENGINES + 1)]

# Dictionary to store the maintenance schedule
maintenance_schedule = {}

# Function to assign maintenance to an engine by a team on a specific day
def assign_maintenance(engine_index, day, team):
    if engine_index not in maintenance_schedule:
        maintenance_schedule[engine_index] = {}
    if day not in maintenance_schedule[engine_index]:
        maintenance_schedule[engine_index][day] = team
    else:
        # Handle case where another team is already assigned for the day
        print("Error: Another team is already assigned for this day.")

# Function to get the maintenance team assigned to an engine on a specific day
def get_maintenance_team(engine_index, day):
    if engine_index in maintenance_schedule and day in maintenance_schedule[engine_index]:
        return maintenance_schedule[engine_index][day]
    else:
        return None


# Example usage
assign_maintenance(1, 1, "T1")
safety_due_date = 29  # Example RUL for engine 1 and Team 1
print(calculate_penalty_cost(1, safety_due_date))

## Fitness evaluation
For each individual in the population we now calculate the fitness value

In our case it is Teams of type A need μAj days to perform maintenance on engine j ∈ M * cost function?
So then we keep iteraitng and we end up with the teams allocated to certain engines?

In [ ]:
# Function to calculate penalty cost for unmaintained engine/machine
def calculate_penalty_cost(engine_index, safety_due_date):
    print(f"Safety due date {safety_due_date} and planning horizon {PLANNING_HORIZON}")
    if safety_due_date >= PLANNING_HORIZON:
        return 0  # No penalty if maintenance is performed within the planning horizon
    else:
        days_past_due = PLANNING_HORIZON - safety_due_date
        daily_cost = engine_costs[engine_index - 1] * (days_past_due ** 2)
        return min(daily_cost, MAX_DAILY_COST)  # Cap the daily cost at MAX_DAILY_COST


## Check termination criteria

### As presented on the slide it is very often unknown, often some threshold is introduced. In our case it is that company X wants to allocate teams to engines that have a predicted safety due date of less than 30

### Termination criteria than is that all engines have due date less than 30?


## Selection of parents

### In our case fitness value is RUL value or penalty cost?

### I am still not sure how many parents we need to choose?

## Crossover

## Mutation (If any)

## New offspring